[JSC] Conform `String.prototype.substring` to TC39 spec

Gumichocopengin8 opened the pull request:
#### dd3af5604c58cde8de0ed7632419d6f386b1b3d3
<pre>
[JSC] Conform `String.prototype.substring` to TC39 spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=300578">https://bugs.webkit.org/show_bug.cgi?id=300578</a>

Reviewed by Yusuke Suzuki and Darin Adler.

This patch conforms `String.prototype.substring` to TC39 spec [1],
such as using `ToIntegerOrInfinity` instead of `ToNumber`.

This patch introduces no performance regressions.

                                                TipOfTree                   Patched                            Ratio
string-substring-constants                 0.4274+-0.0569            0.4026+-0.0112          might be 1.0617x faster
string-substring-constants-binary          0.3904+-0.0315     ?      0.3952+-0.0442        ? might be 1.0122x slower
string-substring-infinity                  0.5556+-0.0095            0.5314+-0.0269          might be 1.0455x faster
string-substring-length-constant           0.3914+-0.0278     ?      0.4163+-0.0378        ? might be 1.0636x slower
string-substring-nan                       0.4938+-0.0259            0.4894+-0.0293
string-substring-constants-identity        0.3593+-0.0146            0.3423+-0.0323          might be 1.0496x faster

[1]: <a href="https://tc39.es/ecma262/#sec-string.prototype.substring">https://tc39.es/ecma262/#sec-string.prototype.substring</a>

Tests: JSTests/microbenchmarks/string-substring-infinity.js
       JSTests/microbenchmarks/string-substring-nan.js
       JSTests/stress/string-substring.js

* JSTests/microbenchmarks/string-substring-constants-binary.js:
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-constants-identity.js:
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-constants.js:
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-infinity.js: Added.
(shouldBe):
(test):
(i.test):
* JSTests/microbenchmarks/string-substring-length-constant.js:
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-nan.js: Added.
(shouldBe):
(test):
(i.test):
* JSTests/microbenchmarks/string-substring.js:
(shouldBe):
(test1):
* JSTests/stress/string-substring.js: Added.
(shouldBe):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/301783@main">https://commits.webkit.org/301783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/16a0dff3d37fe581c8f02728e5190298a257ad59

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126599 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133573 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78269 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46878 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54782 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96343 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64425 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37504 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113233 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76869 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36394 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31415 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76970 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/118649 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107318 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31709 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136141 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125064 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53290 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40997 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104854 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53776 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109586 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104560 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26746 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50033 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28378 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/50722 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53210 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59023 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158109 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52491 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39564 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55825 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54226 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -934,30 +934,19 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSubstring, (JSGlobalObject* globalObject
-    double start = a0.toNumber(globalObject);
+    double start = a0.toIntegerOrInfinity(globalObject);
     RETURN_IF_EXCEPTION(scope, { });
+    start = std::clamp<double>(start, 0.0, len);
```
We would probably get slightly better performance by doing fewer conversions from `double` to `unsigned` if we did this:
```suggestion
    double startDouble = a0.toIntegerOrInfinity(globalObject);
    RETURN_IF_EXCEPTION(scope, { });
    unsigned start = std::clamp<double>(startDouble, 0.0, len);
```

Gumichocopengin8 replied:
Fixed

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -934,30 +934,19 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSubstring, (JSGlobalObject* globalObject
+    double start = a0.toIntegerOrInfinity(globalObject);
     RETURN_IF_EXCEPTION(scope, { });
+    start = std::clamp<double>(start, 0.0, len);
     double end;
```
And here:
```suggestion
    unsigned end;
```

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -934,30 +934,19 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSubstring, (JSGlobalObject* globalObject
-        double temp = end;
-        end = start;
-        start = temp;
+        end = std::clamp<double>(end, 0.0, len);
```
Same here:
And here:
```suggestion
        double endDouble = a1.toIntegerOrInfinity(globalObject);
        RETURN_IF_EXCEPTION(scope, { });
        end = std::clamp<double>(end, 0.0, len);
```

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -934,30 +934,19 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSubstring, (JSGlobalObject* globalObject
     }
-    unsigned substringStart = static_cast<unsigned>(start);
-    unsigned substringLength = static_cast<unsigned>(end) - substringStart;
+    auto [substringStart, substringEnd] = std::minmax(start, end);
```
If we decide not to do the above, then I think this smaller change might also slightly improve performance by cutting down on the number of conversions from `double` to `unsigned`:
```suggestion
    auto [substringStart, substringEnd] = std::minmax<unsigned>(start, end);
```

darinadler commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -89,12 +89,10 @@ ALWAYS_INLINE JSString* stringSlice(JSGlobalObject* globalObject, VM& vm, JSStri
-    if (endValue)
-        end = std::min<int32_t>(std::max<int32_t>(endValue.value(), 0), length);
+    int32_t start = std::clamp<int32_t>(startValue, 0, length);
+    int32_t end = std::clamp<int32_t>(endValue.value_or(length), 0, length);
```
These seem like style changes that, while harmless, don’t change behavior at all. Maybe we could land this separately?

Gumichocopengin8 replied:
Will make a new PR for it
#52455

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -934,30 +934,19 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncSubstring, (JSGlobalObject* globalObject
     }
-    unsigned substringStart = static_cast<unsigned>(start);
-    unsigned substringLength = static_cast<unsigned>(end) - substringStart;
+    auto [substringStart, substringEnd] = std::minmax<unsigned>(start, end);
```
No need to specify the type here.
```suggestion
    auto [substringStart, substringEnd] = std::minmax(start, end);
```

Gumichocopengin8 replied:
Updated

Gumichocopengin8 committed:
[JSC] Conform `String.prototype.substring` to TC39 spec
https://bugs.webkit.org/show_bug.cgi?id=300578

Reviewed by Yusuke Suzuki and Darin Adler.

This patch conforms `String.prototype.substring` to TC39 spec [1],
such as using `ToIntegerOrInfinity` instead of `ToNumber`.

This patch introduces no performance regressions.

                                                TipOfTree                   Patched                            Ratio
string-substring-constants                 0.4274+-0.0569            0.4026+-0.0112          might be 1.0617x faster
string-substring-constants-binary          0.3904+-0.0315     ?      0.3952+-0.0442        ? might be 1.0122x slower
string-substring-infinity                  0.5556+-0.0095            0.5314+-0.0269          might be 1.0455x faster
string-substring-length-constant           0.3914+-0.0278     ?      0.4163+-0.0378        ? might be 1.0636x slower
string-substring-nan                       0.4938+-0.0259            0.4894+-0.0293
string-substring-constants-identity        0.3593+-0.0146            0.3423+-0.0323          might be 1.0496x faster

[1]: https://tc39.es/ecma262/#sec-string.prototype.substring

Tests: JSTests/microbenchmarks/string-substring-infinity.js
       JSTests/microbenchmarks/string-substring-nan.js
       JSTests/stress/string-substring.js

* JSTests/microbenchmarks/string-substring-constants-binary.js:
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-constants-identity.js:
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-constants.js:
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-infinity.js: Added.
(shouldBe):
(test):
(i.test):
* JSTests/microbenchmarks/string-substring-length-constant.js:
(shouldBe):
(test1):
* JSTests/microbenchmarks/string-substring-nan.js: Added.
(shouldBe):
(test):
(i.test):
* JSTests/microbenchmarks/string-substring.js:
(shouldBe):
(test1):
* JSTests/stress/string-substring.js: Added.
(shouldBe):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/301783@main
